### ISSUE TITLE
cmd/compile: functions with switches cannot be inlined

### ISSUE THREAD

**Issue by dsnet** [golang/go#13071]
Using ```go1.5```

Consider the following package:
```go
package main

type dictDecoder struct {
	hist []byte
	wrPos int
}

func (dd *dictDecoder) LastBytesA() (p1, p2 byte) {
	if dd.wrPos > 1 {
		return dd.hist[dd.wrPos-1], dd.hist[dd.wrPos-2]
	} else if dd.wrPos > 0 {
		return dd.hist[dd.wrPos-1], dd.hist[len(dd.hist)-1]
	} else {
		return dd.hist[len(dd.hist)-1], dd.hist[len(dd.hist)-2]
	}
}

func (dd *dictDecoder) LastBytesB() (p1, p2 byte) {
	switch {
	case dd.wrPos > 1:
		return dd.hist[dd.wrPos-1], dd.hist[dd.wrPos-2]
	case dd.wrPos > 0:
		return dd.hist[dd.wrPos-1], dd.hist[len(dd.hist)-1]
	default:
		return dd.hist[len(dd.hist)-1], dd.hist[len(dd.hist)-2]
	}
}
```

Using ```go build -gcflags=-S dict.go```, I see that LastBytesA and LastBytesB produce the exact same assembly output.

However, ```go build -gcflags=-m dict.go``` reports that only LastByteA can be inlined and not LastBytesB:
```bash
# command-line-arguments
./main4.go:8: can inline (*dictDecoder).LastBytesA
./main4.go:8: (*dictDecoder).LastBytesA dd does not escape
./main4.go:18: (*dictDecoder).LastBytesB dd does not escape
```

I expect both LastBytesA and LastBytesB to be inline-able if they produce the same assembly output. This would be nice since (IMHO), switch statements are cleaner for some if-elseif-elseif-elseif-else blocks.

**Comment by randall77** [comment]
This would be nice as part of a general revisit of inlining decisions.  I don't think this is on anyone's radar at the moment, though.

See #12312 We'll need noinline directives first so we can stop using switch{} as a noinline hint (should happen soon).

**Comment by randall77** [comment]
A bunch of tests until recently used switch{} to disable inlining.  We just added a go:noinline directive to replace these uses.  We'll need to make sure we replace all of the switch{} uses and then we can fix this bug.

**Comment by dsnet** [comment]
That sounds wonderful. Would that also fix #7655? And would this make it into go1.6?

Regardless, I'd be happy to abandon that CL of mine.

**Comment by dsnet** [comment]
@randall77 You mentioned that it might be a good time to revisit inline decisions. Is that taking placing somewhere? I would be interested in being involved in that discussion. I've been working on trying to make compression libraries more performant, and many of the changes I make are related to coercing the compiler to inline certain functions.

**Comment by randall77** [comment]
I know of no one working on inline decisions right now.
This bug and #7655 should be pretty easy to fix.
Any fix would not make 1.6, as we are already in the freeze.
